src/Core/BufferLine.re

```diff
@@ -17,14 +17,17 @@ type characterCacheInfo = {
   width: int,
 };
 
-let emptyArray = Array.make(0, None);
+let emptyArrayCharacters = Array.make(0, None);
+let emptyArrayIndex = Array.make(0, None);
```

Comment:
This confuses me. First of all, arrays are mutable, so a simple mistake could suddenly have these be non-empty. Secondly, why are there two of them, if they're not (supposed to be) mutated? Are their physical identity significant somehow? Why not just use the empty array literal `[||]` and compare structurally against that? Or use `Array.length` to check for emptiness.

I could see one reason for this being that the empty array literal will allocate more that doing this, although I would think the compiler should be able to optimize that away. But if it is, this should be documented with a comment. And it's still quite error-prone because of the assumption that it won't be mutated.

Reply:
Actually, native OCaml doesn't have dynamic arrays. I was thinking in terms of JavaScript/BuckleScript where even empty arrays can be mutated. That shouldn't be a problem here, but this still is very weird code to me, that at least deserves a comment explaining it.

Reply:
I added more description about these empty / sentinel values - hopefully it helps explain how they are used.

Reply:
Thanks. I still think it's weird that the compiler isn't able to optimize away allocation of empty arrays though. Seems like a trivial thing to do, and very low-hanging fruit.

Reply:
I created a benchmark to check - it seems that it does optimize away the allocation, but the structural equality check is still more expensive than a reference equality check:
```
|---------------------------------------------------------------------+-------------+---------------------+------------+------------+---------------+-------------+---------------|
|  Array: structural equality check                                   |  1000000    |  0.0139060020447    |  0         |  0         |  27           |  0          |  0            |
|---------------------------------------------------------------------+-------------+---------------------+------------+------------+---------------+-------------+---------------|
|  Array: reference equality check                                    |  1000000    |  0.00193691253662   |  0         |  0         |  27           |  0          |  0            |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

Reply:
Yeah, that's probably because it's polymorphic, and why Base/Core and others don't have them in `Pervasives`.